spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala

```diff
@@ -855,19 +855,57 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
           None
 
         case EqualTo(left, right) =>
-          val leftExpr = exprToProtoInternal(left, inputs)
-          val rightExpr = exprToProtoInternal(right, inputs)
+          val zero = Literal.default(left.dataType)
+          val negZero = UnaryMinus(zero)
 
-          if (leftExpr.isDefined && rightExpr.isDefined) {
-            val builder = ExprOuterClass.Equal.newBuilder()
-            builder.setLeft(leftExpr.get)
-            builder.setRight(rightExpr.get)
+          def buildEqualExpr(leftExpr: Expr, rightExpr: Expr): ExprOuterClass.Expr = {
+            ExprOuterClass.Expr
+              .newBuilder()
+              .setEq(
+                ExprOuterClass.Equal
+                  .newBuilder()
+                  .setLeft(leftExpr)
+                  .setRight(rightExpr))
+              .build()
+          }
+
+          if (left.dataType == DoubleType || left.dataType == FloatType) {
+            // make sure left or right is not null
+            (left, right) match {
+              case (`negZero`, _) =>
```

Comment:
`case (negZero, negZero)` may fall into this case as well? If so only left side gets `Abs()`?